source/common/common/cleanup.h

```diff
@@ -55,4 +55,42 @@ template <class T> class RaiiListElement {
   bool cancelled_;
 };
 
+// RAII helper class to add an element to a std::list held inside an absl::flat_hash_map on
+// construction and erase it on destruction, unless the cancel method has been called. If the list
+// is empty after removal of the element, the destructor will also remove the list from the map.
+template <class Key, class Value> class RaiiMapOfListElement {
+public:
+  using MapOfList = absl::flat_hash_map<Key, std::list<Value>>;
+
+  template <typename ConvertibleToKey>
+  RaiiMapOfListElement(MapOfList& map, const ConvertibleToKey& key, Value value)
+      : map_(map), list_(map_.try_emplace(key).first->second), key_(key), cancelled_(false) {
+    it_ = list_.emplace(list_.begin(), value);
+  }
+
+  virtual ~RaiiMapOfListElement() {
+    if (!cancelled_) {
+      erase();
+    }
+  }
+
+  void cancel() { cancelled_ = true; }
+
+  void erase() {
```

Comment:
I don't think you use this directly right now? Either make this private or inline it above?

Reply:
Fixed in 3e16973